### ISSUE TITLE
Sort by date does not does not correctly handle timezones

### ISSUE THREAD

**Issue by mcdamo** [bpatrik/pigallery2#871]
Sort by date does not does not correctly handle timezones.

In the case that a directory or search contains media with and without timezone data they will not display in the expected order.

Cameras generally store the _Local Time_ that the media was taken but modern cameras or smartphones can now record the Timezone offset as well.

The problem arising with sorting in Pigallery appears because all media is converted to UTC during indexing.

For example two photos taken at the same time, midday in GMT+09, one on a camera and the other on a smartphone
  1. on a camera **without timezone data** will be indexed as `12:00` with `0` offset
  2. on a smartphone **with timezone** will be indexed as `03:00` with `+09:00` offset. 

This means that the first photo will be sorted as if it were 9 hours later than the second within Pigallery.

Here in the code the sorting services uses the base UTC time without referencing the offset:
https://github.com/bpatrik/pigallery2/blob/6c53aca7e7d5b65d257c80737647d57854592bbe/src/frontend/app/ui/gallery/navigator/sorting.service.ts#L153

It may be more coherent to sort all media by the _Local Time_ instead of UTC.

**Comment by bpatrik** [comment]
Thank you for the report.

@grasdk do you have a quick insight to this issue?

**Comment by grasdk** [comment]
@bpatrik: I think @mcdamo already found the problem (or part of it).

It appears that I missed `sorting.service.ts`, when adding comments about searching and filetering should take offset into account, when I introduced offset to the metadata. 

I wrote some [//TODOs in the searchmanager](https://github.com/bpatrik/pigallery2/pull/829/files#diff-f85ded2f38ddf78f5cb870136c6928fe28a9532edf5ceaaee365c8120ad8ec29R566-R571) and in the [filter.service.ts](https://github.com/bpatrik/pigallery2/pull/829/files#diff-72075ba83c7886c7221fdcbdcd5912a9d95926d5c423f3de85aeb5f7cf93c333R208), but not in `sorting.service.ts`.

Basically, one can do something like

`return (a.metadata.creationDate + Utils.getOffsetMinutes(a.metadata.creationDateOffset)*60000) - (b.metadata.creationDate + Utils.getOffsetMinutes(b.metadata.creationDateOffset)*60000)`

in the line that @mcdamo writes.

This will take the offset as ms (minutes*60000) and add it to the creationDates before comparing. Addition should work fine with negative offsets. There are some edge cases, where either or both offsets are not defined, that needs to be handled, in or before the line I suggested.

**Comment by grasdk** [comment]
Hey @mcdamo and @bpatrik

I had a look at this today. I guess the real solution here is a configuration, perhaps an omnipresent ui element that will allow you to switch between: _picture's local time_ and _fixed timezone_ (and _fixed timezone_ can be configured in the config to UTC or any timezone you prefer).

Right now it's a mix. The deep down representation of a photo's timestamp is UTC. This is used for sorting in the line you discovered above.

Some months ago I added the offset information to this, so you _can_ calculate the picture's local time. This is shown in the info of pictures and it seems to be in effect in the list view of images as well. So I guess it's a bit inconsistent and I can see that in some cases, you would like sorting to function on picture's local time instead.

- @mcdamo: For your particular case, as you also write. Your incomplete metadata causes the problem with sorting. One camera stores no timezone offset and Pigallery2 can't guess that. I would probably fix the metadata with `exiftool` or the like. It would also be solved if you had the switch I suggested above.
-- I will try and work on the code to make this possible in the coming time, though I'm not good with all the ui elements. But I could make the functionality ready and perhaps someone else would make the UI stuff (or I could learn) :)

_______________________________________

PS: I'm attaching some pictures I generated for showing the problem. If we pretend they are photos taken all over the world, they are taken within 4 minutes in total, but have local time with great differences. Right now they're sorted in UTC/Global time. It makes sense in this scenario: If 7 friends talk over the Internet, asking each other to take photos of their location and all upload them to the same storage via a service that doesn't ruin the metadata, they would show up in the exact order they were taken! Sorting in local time would change that a lot. :)

Here is a screenshot of how pigallery sorts them now:
![image](https://github.com/bpatrik/pigallery2/assets/115414609/2145705e-6831-4d68-a230-bbc58403faa6)

Here are the pictures - feel free to use:
![UTC-20240413-203430-Auckland](https://github.com/bpatrik/pigallery2/assets/115414609/5fa062a9-0e09-40f8-adad-f681ce255ca3)
![UTC-20240413-203530-Montevideo](https://github.com/bpatrik/pigallery2/assets/115414609/e9f2241e-def7-4ccd-bbb9-71689bedb6dd)
![UTC-20240413-203531-Kathmandu](https://github.com/bpatrik/pigallery2/assets/115414609/1fb16065-318d-4360-bc35-35f659029712)
![UTC-20240413-203630-London](https://github.com/bpatrik/pigallery2/assets/115414609/bf45e0ca-06a3-429b-b2e2-8c9606771eed)
![UTC-20240413-203720-Tokyo](https://github.com/bpatrik/pigallery2/assets/115414609/31022c8a-0d5e-47ab-9404-874b5757f549)
![UTC-20240413-203730-Reykjavik](https://github.com/bpatrik/pigallery2/assets/115414609/fd9ed5f9-6a35-4040-ac88-33e6a80b789a)
![UTC-20240413-203830-NewYork](https://github.com/bpatrik/pigallery2/assets/115414609/a01c1865-8562-4143-a101-67e1fc15e668)

**Comment by mcdamo** [comment]
@grasdk  I think the example of multiple people from different timezones submitting photos to a single album is probably not a common use case for pigallery?

Deep down in the EXIF data a photo's timestamp is a date string representing when the photo was taken. I believe this is _commonly_ local-time that the camera is configured in, though the spec is ambiguous.

The timezone offset was not added to the [EXIF spec](https://en.wikipedia.org/wiki/Exif#Time_Tags) until v2.31. While this spec was reportedly released in mid 2016, it was many years after that manufacturers started using.

For a real world example a smartphone I tested added the offset tag via software update in 2019, while a 2021-camera I tested did not initially support the offset tag and later added it via firmware update in 2023. Earlier make cameras simply do not use this offset tag.

The issue with pigallery converting all image metadata to UTC is that anyone using **any** images or videos without offset tags this is going to cause ongoing confusion and problems.

I would prefer to completely ignore the timezone offset every where that metadata is used, including searching, filtering, and sorting. In effect always using the photo's recorded time and **never** converting between timezones.

**Comment by grasdk** [comment]
A fair point and opinion. I think we must agree to disagree on what 'common' is.

I'm looking at offset as an important piece of data that should be taken into account. Of course if your pictures are not homogeneous (all having offset or all missing offset), this will lead to problems. 

Personally, when I get pictures that lack any tag that I'm interested in (tags, timestamps, offsets, toponomy names, gps), I manually add them to the picture metadata rather than expecting tools to "guess it". Geosetter, DigiKam and exiftool are among the tools I use for this.

However, everyone is not as rigid in metadata homogeneity, so I will work on making a configuration for PiGallery so you can configure it to ignore offset. This will be my next focus for contribution to PiGallery. :-)

**Comment by bpatrik** [comment]
@mcdamo 
> @grasdk I think the example of multiple people from different timezones submitting photos to a single album is probably not a common use case for pigallery?

One would think so, but I'm pretty sure that if we make this assumption, issues will appear like "I'm using pigallery with a single folder. I love it, but now its broken". I have seen this before.

On the broader topic: I agree that there should be a clear way how pigalley handles time. (this is not a simple problem actually. timezones are hard). I think having now supporting offsets is a way forward. 
I think a general user experience should be that user can see their photos in the same circumstance as it was taken. Eg: if it was taken in NY noon, using the app in London, the app should still show noon. 
Being able to see the photo with relative time to our current timezone also has value. Especially if timezones are mixing in one folder. 
Search, filtering, sorting should also "as user expects"

I think use should be able to switch between different timezone settings. (as @grasdk mentions in prev. comment)

**Comment by grasdk** [comment]
@bpatrik @mcdamo 

I've been a bit away from this, but I have thought about it. Yesterday I began looking for references to metadata creation date, because basically everywhere the data is used, we need to be in control.
So there are quite a lot of references to this data outside the metadataloader itself.

My my take would be a multi stage refactor:

1. No functional change, not resolving this task or any other, but lining up the code to do something about it: 
Replace any reference to creation data with a function call that includes creation date and offset and returns a tuple (list) of date and offset. The function will just return the unaltered values of creation date and offset and not take offset into account. Why? To line up the code with a central place where changes can be made.
2. Extend the function with the ability to read configuration-like values:
1: Fallback_offset: A value that will be used as offset, when offset is not present. Today it defaults to the local timezone of the server. Default value Empty (whatever the default is for JS). Typical value: Your personal timezone or maybe UTC if you're into such things ;). The choice is the user's.
2: Override with fallback_offset: A value that if true is similar to ignoring offset alltogether. But to be strict, it is more precise to override, than to ignore.
3. Create configuration options for Fallback_offset and Override with fallback_offset. Perhaps "Override with fallback_offset" should be an omnipresent button than you can enable and disable anywhere. And it _could_ be named  "Ignore timezone" and maybe we can find or create a nice icon for it.
4. Make the function in 2 use the solution of 3. :)

what do you think? is there something obvious I have overlooked. I know it's laid out simply and things like database searches etc. need special attention.


a boolean that decides "ignore offset" and makes sure that the input timestamp is returned without change, and the offset is returned as 


-

**Comment by grasdk** [comment]
@mcdamo I would say this bug is resolved in the latest edge version, if you leave the timestamp offset setting on "ignore". See https://github.com/bpatrik/pigallery2/pull/902

**Comment by mcdamo** [comment]
Thanks @grasdk, this option resolves the sorting of images without timezones.
